packages/quantification/src/net-quantification.ts

```diff
@@ -0,0 +1,185 @@
+import type {
+  AnnualTotals,
+  UnadjustedQuantificationSummary,
+} from './quantification';
+
+/**
+ * Calculates the net tonnes removed accounting for years with emissions. Years with emissions (represented
+ * by a negative number) should be reduced to zero and any tonnes removed should be reduced by the emitted
+ * amount.
+ *
+ * Algorithm overview:
+ * 1. Years with negative numbers are set to zero
+ * 2. The amount emitted will be debited from the subsequent years of the same field
+ * 3. Should the amount emitted exceed the amount available from subsequent years, wrap to the
+ * start year and continue to the currently considered year
+ * 4. Should the amount exceed all available tonnes removed from this project, repeat step 1
+ * starting with the currently considered year in the next quantification
+ * 5. Should the negative amount exceed the total tonnes available, the negative amount should
+ * persist in the year it originated
+ *
+ * An example:
+ *
+ * Starting with:
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 10   | -30  | 9    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 30 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 10   | 0    | 9    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 21 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 10   | 0    | 0    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 11 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 0    | 0    | 0    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 1 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 0    | 0    | 0    |
+ * | B     | 10   | 0    | 9    |
+ *
+ * 0
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 0    | 0    | 0    |
+ * | B     | 10   | 0    | 8    |
+ * | final | 10   | 0    | 8    |
+ *
+ */
+export const getNetQuantificationProjection = (
+  quantifications: Pick<
+    UnadjustedQuantificationSummary,
+    'somscAnnualDifferencesBetweenFutureAndBaselineScenarios'
+  >[],
+  logger?: Pick<Console, 'debug' | 'table'>
+): AnnualTotals[] => {
+  const netQuantifications: AnnualTotals[] = quantifications.map(
+    (quantification) => ({
+      ...quantification.somscAnnualDifferencesBetweenFutureAndBaselineScenarios,
```

Comment:
Could probably just use

```ts
Object.fromEntries(
 Object.entries(
   quantification.somscAnnualDifferencesBetweenFutureAndBaselineScenarios
 )
 .sort() // should correctly sort the years (tested with [{'2029':1,'2020':2,'2025':undefined}].sort()[0])
 .map(([k, v]) => [k, v ?? 0]) // assigns any `undefined` annual data to 0
)
```
which would allow you to skip the next ~20 lines of code

Reply:
Sorry, I'm not sure that works, but maybe I'm not seeing it. The next loop on line 76 gathers years from all quantifications. For example:

```ts
const a = { 2010: 5, 2011: 5 }
const b = { 2011: 5, 2012: 5 }

// we want the set of years to be:
[ 2010, 2011, 2012 ]
```

This evens out our row-column traversal and gives us an even grid. We do need the ordered set but javascript made some weak guarantees about key order  in objects. I'd rather be confident and keep the years in an array. 

Further down, we add the year 0 to every undefined year key, but we need the full set of years first before we can do that.

Am I missing something here? Feel free to bring me in to a call if you want to discuss.

Reply:
Ah I missed that requirement

How about

```ts
c.reduce(
  (accumulator, next) => [
    ...accumulator,
    Object.assign(
      ...[...new Set(c.flatMap(Object.keys))].map((k) => ({ [k]: 0 })),
      next
    ),
  ],
  []
);
```

Reply:
![image](https://user-images.githubusercontent.com/18407013/171949083-0179c822-08ba-45f1-89f1-005780d10adc.png)